lib/google/gax/bundling.rb

```diff
@@ -0,0 +1,448 @@
+                            bundling_request,
+                            subresponse_field: bundle_desc.subresponse_field)
+          delay_threshold_millis = @options.delay_threshold_millis
+          if delay_threshold_millis > 0
+            run_later(bundle.bundle_id, delay_threshold_millis)
+          end
+          @tasks[bundle_id] = bundle
+          return bundle
+        end
+      end
+
+      # Helper function for #schedule.
+      #
+      # Creates a new thread that will execute the encapsulated api calls after
+      # the +delay_threshold_millis+ has elapsed. The thread that is
+      # spawned is added to the @threads hash to ensure that the thread will
+      # api call is made before the main thread exits.
+      #
+      # @param bundle_id [String] the id corresponding to the bundle that
+      #     is run.
+      # @param delay_threshold_millis [Numeric] the number of micro-seconds to
+      #     wait before running the bundle.
+      def run_later(bundle_id, delay_threshold_millis)
+        Thread.new do
+          @timer.run_after(delay_threshold_millis / MILLIS_PER_SECOND) do
+            run_now(bundle_id)
+          end
+        end
+      end
+
+      # Helper function for #schedule.
+      #
+      # Immediately runs the bundle corresponding to the bundle id.
+      # @param bundle_id [String] the id corresponding to the bundle that
+      #     is run.
+      def run_now(bundle_id)
+        @tasks_lock.synchronize do
+          if @tasks.key?(bundle_id)
+            task = @tasks.delete(bundle_id)
+            task.run
+          end
+        end
+      end
+
+      # This function should be called before the main thread exits in order to
+      # ensure that all api calls are made.
+      def close
+        @tasks_lock.synchronize do
+          @tasks.each do |bundle_id, _|
+            run_now(bundle_id)
+          end
+        end
+      end
+
+      private :bundle_for, :run_later, :run_now
+    end
+
+    # Container for a thread adding the ability to cancel, check if set, and
+    # get the result of the thread.
+    class Event
+      attr_accessor :canceller
+      attr_reader :result
+
+      def initialize
+        @canceller = nil
+        @result = nil
+        @is_set = false
+        @mutex = Mutex.new
+        @resource = ConditionVariable.new
+      end
+
+      def result=(obj)
+        @mutex.synchronize do
+          @result = obj
+        end
+      end
+
+      # Checks to see if the event has been set. A set Event signals that
+      # there is data in @result.
+      # @return [Boolean] Whether the event has been set.
+      def set?
+        @is_set
+      end
+
+      # Signifies that the event has been set and that there is data in @result.
+      def set
+        @mutex.synchronize do
+          @is_set = true
+          @resource.broadcast
+        end
+      end
+
+      # Resets the Event to not being set and clears the @result.
+      def clear
```

Comment:
Not sure what is the purpose of this method. Gax-Python has this because Python's threading.Event has a method like this, please clarify why it needs to be in Ruby too.